### ISSUE TITLE
Slurm adapter: srun cannot see $PATH/other env vars when submitted through job composer

### ISSUE THREAD

**Issue by jasonbuechler** [OSC/ood_core#158]
Let me start off by stating I'm honestly not sure if this is an OOD issue or a Slurm issue. I'm starting but submitting the issue with OOD because everything still works fine outside of Job Composer. We simultaneously upgraded both Slurm (18 to 19) and OOD (to 1.6.20/1.35.3), so it's tough to say for certain, but everything was working fine before the upgrades.

In a nutshell: using job-composer (and only using job-composer) to submit a batch script will change the behavior of `srun ...` lines within the script: programs executed from srun have a very different set of environment variables (and no $PATH whatsoever) than programs/commands within the same script that are not prefixed with srun.  A workaround is to always explicitly use `srun --export=all ...`  but this is not necessary when `sbatch`ing from any command line (including OOD's builtin terminal) and also seems contrary to slurm's documentation for srun:

> –export=<environment variables [ALL] | NONE>
Identify which environment variables are propagated to the launched application. By default, all are propagated.

This is probably best demonstrated by example, so take a peek at the below output taken from running the attached batch file through Job Composer:

```
OnDemand version: v1.6.20 | Dashboard version: v1.35.3
Slurm version: slurm 19.05.3-2


EXPECTATION: srun lines have same access to environment
             vars (esp PATH!) as lines not using srun.

** module load perl    # or any binary in PATH ...

** /usr/bin/which perl ...
/packages/perl/5.28.1/bin/perl

** srun --export=ALL /usr/bin/which perl ...
/packages/perl/5.28.1/bin/perl

** srun /usr/bin/which perl ...
/usr/bin/which: no perl in ((null))
srun: error: cn53: task 0: Exited with exit code 1
#         ^---- REALITY: non-srun command works as expected...
#                        BUT the srun command fails.

EXPECTATION: srun defaults to --export=ALL behavior as
             is stated in the srun docs/man-page.

** /bin/env | grep -vc SLURM     # count non-slurm env vars ...
** /bin/env | grep ^PATH | grep -o ':' | wc -l    # count PATH elements ...
35
10

** srun --export=ALL /bin/env | grep -vc SLURM ...
** srun --export=ALL /bin/env | grep ^PATH | grep -o ':' | wc -l ...
40
10

** srun /bin/env | grep -vc SLURM ...
** srun /bin/env | grep ^PATH | grep -o ':' | wc -l ...
6
0
# ^---- REALITY: non-srun command works as expected...
#                BUT the srun command does not reflect the greater sbatch job-env.
```

This all seems pretty related to issue #109 and this thread https://listsprd.osu.edu/pipermail/ood-users/2018-December/thread.html#329 ... but not closely enough for me to immediately make use of either for a solution.

sbatch source:
[ood-srun-example.txt](https://github.com/OSC/ood_core/files/3810640/ood-srun-example.txt)

**Comment by ericfranz** [comment]
When NGINX is started by root, the environment is stripped. So if you depend on special variables in the user's environment being set prior to [executing sbatch](https://github.com/OSC/ood_core/blob/a6bc6b28301bd58527dfc439026f0f15e8ffba8e/lib/ood_core/job/adapters/slurm.rb#L196) - we actually use sbatch instead of srun.

What OnDemand version did you install from? We upgraded both Passenger and NGINX in the 1.4 release: https://osc.github.io/ood-documentation/master/release-notes/v1.4-release-notes.html#upgrade-to-ruby-2-4-nodejs-6-passenger-5. In the 1.4.4 changelog https://github.com/OSC/ondemand/blob/master/CHANGELOG.md#144---2018-12-04 there is this:

>Since NGINX 1.14 now strips environment, we explicitly pass these env vars to NGINX: PATH LD_LIBRARY_PATH X_SCLS MANPATH PCP_DIR PERL5LIB PKG_CONFIG_PATH PYTHONPATH XDG_DATA_DIRS SCLS

Those are in the environment of root and can be expanded by adding to /etc/ood/profile (sourced by root). Though you also have to declare any new env vars to pass through to the PUN in the nginx_stage config https://github.com/OSC/ondemand/blob/6d0a707e30fcf74b94767245dfacbbc5c5a8c833/nginx_stage/share/nginx_stage_example.yml#L34-L43

**Comment by ericfranz** [comment]
This is why I [recommend in the Discourse topic](https://discourse.osc.edu/t/ondemand-with-slurm-based-sytems-sbatch/581/8?u=efranz) to use the bin_overrides and a wrapper script for sbatch that is executed in an ssh session.

**Comment by jasonbuechler** [comment]
Hi @ericfranz thanks for bearing with me on this! I'm sorry to ask in advance that you bear with for one more brainstorming session, please and thanks!

The we upgraded OOD from v1.5-1 which we'd been using since 04Mar2019, when we had upgraded from the original install of v1.3-1.  So given that we weren't experiencing the issue with v1.5 I feel like it's safe to conclude that the root issue doesn't involve the nginx change from the 1.4 release.  (But correct me if need be!)

I also just slightly updated the codeblock above, and wanted to draw your attention to the lines where srun apparently fails to return results consistent with the "rest" of the sbatch environment... 
Within the bounds of the exact same sbatch script, *other* commands (otherwise identical, but *omitting the `srun ...` prefix*), still do return results that I am expecting.  
So in this specific case, would engineering the sbatch wrapper approach actually have an effect?   


Now, it's entirely possible that I'm naively expecting something silly here, due to missing some crucial piece of the workflow here, so if that's the case, please allow me to lay out my understanding:

1. OOD calls `sbatch` using, at least, the user-specified sbatch script 
2. the sbatch environment now has environment variables consistent with those explicitly passed (as you listed above)
3. the env vars, including $PATH can be modified by the script as necessary now (e.g. `module load ...`)
4. subsequent script lines can read/interact with these env vars (e.g. run a binary that was not previously in the $PATH)
5. srun can also read/interact with these env vars "in the same way"
6. as per the slurm documentation for `srun`, the default "behavior" (e.g. interactivity with env vars) should be identical to the behavior achieved by using `srun --export=ALL`

...this is accurate to within a 5-minute summary, right?  Or isssss it?

**Comment by ericfranz** [comment]
Your understanding is accurate with the clarification that sbatch is called with [`"SBATCH_EXPORT" => "NONE"` set by default](https://github.com/OSC/ood_core/blob/8d61d9470165b12413fa66bd208ac865db8bc45f/lib/ood_core/job/adapters/slurm.rb#L195)

Also, I do not see any [changes from 1.5.1 to 1.6.19](https://github.com/OSC/ondemand/compare/v1.5.1...v1.6.19) that would have caused the problem you are facing. Perhaps it is a Slurm issue?

**Comment by ericfranz** [comment]
I think the benefit of a wrapper is that the ssh session has a fresh environment that should be close to identical to the one a user would have when they ssh to that node. Which is definitely not the case for the Per User NGINX.

**Comment by ericfranz** [comment]
Talked offline. Issue is `"SBATCH_EXPORT" => "NONE"` is problematic in certain cases and being able to make the default `"SBATCH_EXPORT" => "ALL"` instead would be preferred. What is the best way to do that in OnDemand, and if not what needs to change to support that? (TODO)

**Comment by yfoo** [comment]
Thanks Eric, looking forward to a solution for this! :)

Chris

**Comment by ericfranz** [comment]
So it looks like in the code there are several problems with setting the environment.

https://github.com/OSC/ood_core/blob/8d61d9470165b12413fa66bd208ac865db8bc45f/lib/ood_core/job/adapters/slurm.rb#L195 sets `"SBATCH_EXPORT" => "NONE"`, which we want to change, but then passes the `OodCore::Job::Script#job_environment` to `execve`, which ironically should be ignored since `SBATCH_EXPORT` is set to `NONE` unless these are specified using the `--export` argument. And they are https://github.com/OSC/ood_core/blob/8d61d9470165b12413fa66bd208ac865db8bc45f. But that only specifies a list of keys of the environment variables to export. And since neither `NONE` nor `ALL` prefix this list of env vars in `--export=` it acts like `--export=NONE,` was specified.

This means the current behavior is:

1. If no custom environment is specified when submitting the job, `--export` argument is not used and `SBATCH_EXPORT=NONE` ensures the environment is not passed through
2. If custom environment is specified, `--export` flag is used, with a list of env vars, but not prefixing `NONE`, meaning that `NONE` is assumed and then later setting `SBATCH_EXPORT=NONE` is redundant. The custom environment is passed to `execve` and thus passed through to the job, since the env vars are specified in the `--export` flag.

So what we want to do is:

1. preserve the current behavior, but only use `--export` to reduce duplication
2. provide the ability, perhaps through the cluster config, to always use `ALL` i.e. `--export=ALL,`

There is actually an abstraction for this that works across multiple adapters. See https://slurm.schedmd.com/rosetta.pdf. They call it "Copy Environment". In Slurm its ALL or specific using `--export=`. In Torque its -V for all or -v with specific or default NONE.  SGE uses -V. LSF doesn't have it though. So we could add a `OodCore::Job::Script#copy_environment` attribute.

Then we should enable setting default values for `OodCore::Job::Script` attributes in the cluster config. For example:

```yaml
v2:
  job:
    adapter: slurm
    script:
      copy_environment: true
```

**Comment by MorganRodgers** [comment]
LSF 10.x appears to support the same syntax as Slurm: https://www.ibm.com/support/knowledgecenter/en/SSWRJV_10.1.0/lsf_command_ref/bsub.env.1.html

**Comment by MorganRodgers** [comment]
In light of `bsub` behaving differently than `squeue` it seems like the correct behavior for `Script#copy_environment` is to return `nil` and allow each adapter to perform their default behavior, and only change that behavior if `#copy_environment` is actually set.

**Comment by jasonbuechler** [comment]
@MorganRodgers thank you both so much for looking into that! 
In the meantime before the merge makes its way into a release -- for our cluster's specific OOD instance, is there any "minor" patch/workaround we could implement so as to remove the need to explicitly add the export flag to our users' srun statements? (Locally applying this pull is just a liiittle more tinkering than I'm comfortable with.)
Thanks again!

**Comment by MorganRodgers** [comment]
You're welcome Jason. If you want to patch in this behavior now then have a sudoer create the following two files:

- `/etc/ood/config/apps/myjobs/initializers/squeue_export_all.rb`
- `/etc/ood/config/apps/dashboard/initializers/squeue_export_all.rb`

with the following content for both files:

```ruby
require "ood_core"

module OodCore
    module Job
        class Script
            def job_environment
                {'ALL' => '1', 'SBATCH_EXPORT' => 'ALL'}.merge(@job_environment || {})
            end
        end
    end
end
```

**Comment by jasonbuechler** [comment]
@MorganRodgers sweet that was dead simple and seemed to work as expected, but with one hiccup: we get the following error at the top of the output file:
`init.c(379):ERROR:109: Unknown shell type 'nologin'`

This isn't familiar to me as a site issue (though I'm not excluding the possibility) so I'm hoping you can say whether or not that's unrelated to the initializers?
thank you!

**Comment by ericfranz** [comment]
Is this experienced from a job submitted through the Job Composer or an a job to start an interactive GUI app submitted through the dashboard?

I wonder if this is due to a missing shebang?

The example job templates and ood_core is currently omitting the shebang from the job scripts that are created. For the interactive apps you could add the shebang by modifying the cluster config in the `batch_connect:` section for `basic:` and `vnc:` add a `header: "#!/bin/bash` line i.e.

```yaml
  batch_connect:
    basic:
      header: "#!/usr/bin/env bash\n"
    vnc:
      header: "#!/usr/bin/env bash\n"
```

Or

```yaml
  batch_connect:
    basic:
      header: "#!/usr/bin/env bash -l\n"
    vnc:
      header: "#!/usr/bin/env bash -l\n"
```

or

**Comment by ericfranz** [comment]
@jasonbuechler check on GitHub for the comment I just added, as I accidentally added the comment before finishing my edits so the email you received won't have the final version!

**Comment by jasonbuechler** [comment]
@ericfranz Sorry for my delay in responding to this, but I just did a bunch of additional testing to make sure I knew the scope of what was going on, and have a few notes to report:

- if I didn't state so before: we're using environment modules v3.2.10 (2012-12-21) and not lmod 
- we see the "init.c(379):ERROR:109: Unknown shell type 'nologin'" error for *any* invocation of 'module' ... e.g. load, purge, --version, etc
- previously, I had incorrectly concluded that `module` commands were still working, even though they were throwing that error: however `module` is **not** functional at all.
- this issue is (as far as I know) specific to jobs submitted through the job composer when utilizing [Morgan's suggested patch](https://github.com/OSC/ood_core/issues/158#issuecomment-564726137)
- my test scripts have all been using a normal bash shebang ... I'm not sure if it's interesting or not to note this, but adding the `-l` flag to the shebang produces the same output BUT with one additional instance of the error, at the very top of the output file.

I'm pretty new to environment modules so I'm not sure where to begin debugging even though this seems the appropriate place to be doing it -- so I'm happy to entertain any testing anyone thinks might be helpful!

Thank you!  --jason

**Comment by ericfranz** [comment]
I'm going to try to reproduce this problem on another Slurm system. I will keep you posted.

**Comment by ericfranz** [comment]
Unfortunately, have not yet successfully reproduced. On another Slurm system, ran with this script using OnDemand v1.6.20 and slurm 18.08.8. (that isn't slurm 19.05.3-2 though...)

My job script was:

```
#!/bin/bash
#
#SBATCH --job-name=test
#SBATCH --output=res.txt
#
#SBATCH --ntasks=1
#SBATCH --time=10:00

echo "module load perl and which perl"
module load perl/5.18.4
/usr/bin/which perl
srun --export=ALL /usr/bin/which perl

echo "count non-slurm env vars"
/bin/env | grep -vc SLURM
echo "count PATH elements"
/bin/env | grep ^PATH | grep -o ':' | wc -l 

echo "srun --export=ALL: count non-slurm env vars"
srun --export=ALL /bin/env | grep -vc SLURM
echo "srun --export=ALL: count PATH elements"
srun --export=ALL /bin/env | grep ^PATH | grep -o ':' | wc -l


echo "srun /bin/env | grep -vc SLURM"
srun /bin/env | grep -vc SLURM
echo "srun /bin/env | grep ^PATH | grep -o ':' | wc -l"
srun /bin/env | grep ^PATH | grep -o ':' | wc -l
```

and output was

```
module load perl and which perl
/opt/packages/perl/5.18.4-nothreads/bin/perl
/opt/packages/perl/5.18.4-nothreads/bin/perl
count non-slurm env vars
61
count PATH elements
28
srun --export=ALL: count non-slurm env vars
62
srun --export=ALL: count PATH elements
28
srun /bin/env | grep -vc SLURM
62
srun /bin/env | grep ^PATH | grep -o ':' | wc -l
28
```

**Comment by jasonbuechler** [comment]
> My job script was:
> 
> ```
> <snip>
> module load perl/5.18.4
> /usr/bin/which perl
> srun --export=ALL /usr/bin/which perl
> <snip>
> ```

Hey @ericfranz thanks for taking that crack at it!  I think I might have spotted a step you'd need to do to replicate... I could be wrong but based on the way you're using srun, I'm thinking you maybe haven't applied [Morgan's patch](https://github.com/OSC/ood_core/issues/158#issuecomment-564726137), to get around the need to use either `--export=ALL` or `/full/path/to/binary`.  

If you actually were using Morgan's patch (I probably use the term too loosely... and should say "shim/bodge"), and it's just my site, it's still not a huge deal considering I'd guess a new release with the "real patch" is probably not far away. If I may ask, could you offer a very very rough ballpark guess on when?

Thanks again!

**Comment by ericfranz** [comment]
Do you have a local test deployment of OnDemand you'd be willing to try out a pre-release with the fix? If so I can provide directions on how to install. It basically amounts to installing from https://yum.osc.edu/ondemand/latest/ instead of https://yum.osc.edu/ondemand/1.6/. Note if you are able, I'll let you know when a release in latest contains the copy_environment feature. We still need to make a release containing this.

As for a new stable release of 1.7, current target is the second week of February.

**Comment by jasonbuechler** [comment]
Hey @ericfranz yes indeed we now have a recent clone of our ood VM ready for dev purposes and I'd love to try out a pre-release with that patching, thank you very much!
(A quick browse of `latest` suggests there hasn't been a release with copy_environment yet, is that accurate?)

**Comment by ericfranz** [comment]
I'll post here when a new release in latest has the copy_environment modifications.

**Comment by jasonbuechler** [comment]
Hi @ericfranz and @MorganRodgers I just dropped in to say I am know successfully using [Morgan's mini-patch](https://github.com/OSC/ood_core/issues/158#issuecomment-564726137), which makes me very (very very) happy and also solves the 'module' mystery I mentioned a few weeks ago.

The module-breaking issue was on our side, and was due to the fact we hotwired (in bash) the `module` function to do some extra usage-tracking stuff.  That "hotwiring" worked fine in the previous OOD version, but the mini-patch must not copy some exotic but necessary part of the environment. 
(And I would wager that https://github.com/OSC/ood_core/pull/163 will probably return our ability to "hotwire", but either way, our OOD is running beautifully again.)

**Comment by johrstrom** [comment]
This should be fixed and resolved in ood_core `0.11.4` which is released in Open OnDemand `1.7.14`.

To get `--export=ALL` add `copy_environment: true` in the `v2.job` portion of the cluster's configuration. 

To get get `--export=ALL,FOO` add `copy_environment: true` in the `v2.job` portion of the cluster's configuration. And add this to your job's `submit.yml.erb`. 
```yaml
script:
  job_environment:
    FOO: BAR
```

The default is (and has always been with the exception of a regression) `--export=NONE`. That actually seems to be OK because the environment `sbatch` is run from is the OOD PUNs' environment which is limited and setup to run the PUN, not arbitrary executables. If you do `--export=ALL` you may have to reload files in `/etc/profile` to get modules or similar bash functionality. 

Hope that helps! I'm closing this now as it seems to be concluded and released. Thanks for the ticket and feel free to reach back out with any other issues!
